services/device-registry/src/test/java/org/eclipse/hono/deviceregistry/FileBasedCredentialsServiceTest.java

```diff
@@ -229,6 +229,40 @@ public void testDoStartLoadsCredentials(final TestContext ctx) {
         assertRegistered(svc, Constants.DEFAULT_TENANT, "sensor1", CredentialsConstants.SECRETS_TYPE_HASHED_PASSWORD, ctx);
     }
 
+    /**
+     * Verifies that credentials are ignored if the startEmpty property is set.
+     *
+     * @param ctx The test context.
+     */
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    @Test
+    public void testDoStartIgnoreCredentialIfStartEmptyIsSet(final TestContext ctx) {
+
+        // GIVEN a service configured with a file name and startEmpty set to true
+        props.setFilename(FILE_NAME);
+        props.setStartEmpty(true);
+        when(fileSystem.existsBlocking(props.getFilename())).thenReturn(Boolean.TRUE);
+        doAnswer(invocation -> {
+            final Buffer data = DeviceRegistryTestUtils.readFile(FILE_NAME);
+            final Handler handler = invocation.getArgument(1);
+            handler.handle(Future.succeededFuture(data));
+            return null;
+        }).when(fileSystem).readFile(eq(props.getFilename()), any(Handler.class));
+
+        // WHEN the service is started
+        final Async startup = ctx.async();
+        final Future<Void> startFuture = Future.future();
+        startFuture.setHandler(ctx.asyncAssertSuccess(s -> {
+            startup.complete();
+        }));
+        svc.doStart(startFuture);
+
+        // THEN the credentials from the file are not loaded
+        startup.await();
+        assertNotRegistered(svc, Constants.DEFAULT_TENANT, "sensor1", CredentialsConstants.SECRETS_TYPE_HASHED_PASSWORD, ctx);
```

Comment:
this is a very weak assertion since it will also be true if the file loaded does not contain such credentials.
instead, you should verify that the *readFile* operation has not been invoked at all. For that, you can omit the `doAnswer` block above and add

```
verify(filesystem, never()).readFile(anyString(), any(Handler.class));
```
down here.

Reply:
oh you can even monitor calls, TIL ! Thanks for the suggestion :smile: